### DISCUSSION TITLE
Regression prediction: determine the revenue based on a graph

### DISCUSSION THREAD

**Discussion by dirkcremers** [pyg-team/pytorch_geometric#7821] [General]
Hi,

I have a question regarding the possibility of predicting the revenue of a graph. The problem is defined as follows:

My main goal involves predicting revenue for a graph, where each node represents a store with its specific characteristics (node features). The stores are interconnected in a fully undirected graph, sharing common information such as possible delivery moments (edge attributes). The data is structured using PyTorch Geometric's Data representation.

`Data(node=node_features, edge_index=edge_index, edge_features=edge_features, y=revenue)`

To tackle this task, I implemented a GNN called GAT (Graph Attention Network) (see model below (note that num_classes is set equal to one)). However, during training, I observed that the model predicts revenue for individual nodes instead of the entire graph, which is my target variable. 

```
class GAT(torch.nn.Module):
    """Graph Attention Network"""
    def __init__(self, num_features, num_classes, hidden_channels, heads=1, edge_dimension=29):
        super().__init__()
        self.gat1 = GATv2Conv(num_features, hidden_channels, heads=heads, edge_dim=edge_dimension)

        self.linear1 = nn.Linear(heads*hidden_channels, hidden_channels)
        self.linear2 = nn.Linear(hidden_channels, hidden_channels)
        self.linear3 = nn.Linear(hidden_channels, hidden_channels)

        self.output = nn.Linear(hidden_channels, num_classes)

    def forward(self, x, edge_index, edge_attr):

        x = self.gat1(x, edge_index, edge_attr=edge_attr)
        x = x.relu()

        x= self.linear1(x)
        x = x.relu()

        x = self.linear2(x)
        x = x.relu()
        x = F.dropout(x, p=0.1, training=self.training)

        x = self.linear3(x)
        x = x.relu()

        x = self.output(x)
        return x
```

One potential solution is to predict revenue for each store (node) and then aggregate these predictions to obtain the graph-level prediction. One simple approach is summing the predicted revenues for all nodes. In this approach, I would predict the revenue for each node, calculate the sum of these values, and then use this sum to determine the loss and apply backpropagation.

Is the solution proposed above the best approach, or are there more effective alternatives? I am open to any tips or suggestions that could enhance the solution.

Thanks!
